gcc/rust/typecheck/rust-tyty-rules.h

```diff
@@ -1087,6 +1092,58 @@ class StrRules : public BaseRules
   StrType *base;
 };
 
+class NeverRules : public BaseRules
+{
+  using Rust::TyTy::BaseRules::visit;
+
+public:
+  NeverRules (NeverType *base) : BaseRules (base), base (base) {}
+
+  virtual void visit (TupleType &type) override { resolved = type.clone (); }
+
+  virtual void visit (ADTType &type) override { resolved = type.clone (); }
+
+  virtual void visit (InferType &type) override { resolved = type.clone (); }
```

Comment:
This case isn't specified in (stable) Rust yet. If `InferType` can't be inferred to anything, it will fall back to the `()` (unit) type when unified with the `NeverType`. This is unstable in Rust and may change to be inferred to the `!` type in the future.

I don't think you have to change anything here yet, but I would add a `FIXME` here, since this will probably cause problems down the road. That is, because if you don't unify it with anything here, it will error after type inference iff no type could be inferred otherwise.[¹] 

Since this is unstable Rust, I don't think you have to address this yet, but I would add a note. Some more explanation about this: https://hackmd.io/@FohtAO04T92wF-8_TVATYQ/SJ0vcjyWL?type=view

---

[¹]: So either you will have to deal with this here in some way (which I don't think you can do?) or mark it here somehow, that it might be the `!` type and implement a fallback after type inference (or however this is down with integer types).

Reply:
Thanks for your review first. I agree with you and I won't deal with this (at least in this PR). Actually when implementing NeverType, I tested what you said with rustc and its behavior really confuses me. I will add a `FIXME` later.

Reply:
I think the confusion comes mainly from `!` getting inferred to `()` currently. But yeah, the never type is just confusing by nature I guess.

Reply:
Exact. I wonder why `!` becomes `()` sometimes. Thanks for the article you provided, which answers the question.